File: phasor.cpp
#include <cmath>
#include <iostream>
#include "phasor.h"

float degree_to_radian(float degrees) {
  return degrees*(M_PI/180);
}

float radian_to_degree(float radians) {
  return radians*(180/M_PI);
}

Number::Number() {
  number_type_ = REAL;
  magnitude_ = 0;
  phase_ = 0;
}

Number::Number(float new_magnitude, float new_phase, number_type type) {
  number_type_ = type;
  magnitude_ = new_magnitude;
  phase_ = new_phase; 
}

float Number::get_magnitude() {
  return magnitude_;
}

float Number::get_phase() {
  return phase_;
}

void Number::set_magnitude(float new_magnitude) {
  magnitude_ = new_magnitude;
}

void Number::set_phase(float new_phase) {
  phase_ = new_phase;
}

Number Number::operator+(Number& input_num) {
  float magnitude_1 = get_magnitude();
  float phase_1 = get_phase();
  float magnitude_2 = input_num.get_magnitude();
  float phase_2 = input_num.get_phase();

  float real_part_1 = magnitude_1 * std::cos(degree_to_radian(phase_1));
  float imag_part_1 = magnitude_1 * std::sin(degree_to_radian(phase_1));
  float real_part_2 = magnitude_2 * std::cos(degree_to_radian(phase_2));
  float imag_part_2 = magnitude_2 * std::sin(degree_to_radian(phase_2));

  float new_magnitude = std::sqrt(pow(real_part_1+real_part_2,2) + pow(imag_part_1+imag_part_2,2));
  float new_phase = std::atan((imag_part_1+imag_part_2)/(real_part_1+real_part_2));

  set_magnitude(new_magnitude);
  set_phase(radian_to_degree(new_phase));
  return *this;
}

Number Number::operator-(Number& input_num) {
  float magnitude_1 = get_magnitude();
  float phase_1 = get_phase();
  float magnitude_2 = input_num.get_magnitude();
  float phase_2 = input_num.get_phase();

  float real_part_1 = magnitude_1 * std::cos(degree_to_radian(phase_1));
  float imag_part_1 = magnitude_1 * std::sin(degree_to_radian(phase_1));
  float real_part_2 = magnitude_2 * std::cos(degree_to_radian(phase_2));
  float imag_part_2 = magnitude_2 * std::sin(degree_to_radian(phase_2));

  float new_magnitude = std::sqrt(pow(real_part_1-real_part_2,2) + pow(imag_part_1-imag_part_2,2));
  float new_phase = std::atan((imag_part_1-imag_part_2)/(real_part_1-real_part_2)); // outputs radians

  set_magnitude(new_magnitude);
  set_phase(radian_to_degree(new_phase));
  return *this;
}

Number Number::operator*(Number& input_num) {
  float magnitude_1 = get_magnitude();
  float magnitude_2 = input_num.get_magnitude();
  float phase_1 = get_phase();
  float phase_2 = input_num.get_phase();

  set_magnitude(magnitude_1 * magnitude_2);
  set_phase(phase_1 + phase_2);
  return *this;
}

Number Number::operator/(Number& input_num) {
  float magnitude_1 = get_magnitude();
  float magnitude_2 = input_num.get_magnitude();
  float phase_1 = get_phase();
  float phase_2 = input_num.get_phase();

  set_magnitude(magnitude_1 / magnitude_2);
  set_phase(phase_1 - phase_2);
  return *this;
}

float get_real(float phase, float magnitude) { // phase in degrees
  float radian_phase = degree_to_radian(phase);
  return magnitude * std::cos(radian_phase);
}

float get_imag(float phase, float magnitude) { // phase in degrees
  float radian_phase = degree_to_radian(phase);
  return magnitude * std::sin(radian_phase);
}

void print_phasor(Number& phasor) {
  std::cout.precision(3);
  const char *angle_char = u8"\u2220";
  std::cout << phasor.get_magnitude() << angle_char << phasor.get_phase();
}

File: phasor.py
import cmath
import math

class phasor:
    def __init__(self, *args): # args should be provided as (magnitude, phase) or (complex_num)
        if(len(args) == 1): # complex number input
            self.magnitude = calc_mag(args[0])
            self.phase = math.fmod(math.radians(calc_phase(args[0])), math.pi)
            self.real = args[0].real
            self.imag = args[0].imag
        elif(len(args) == 2): # magnitude, phase input
            self.magnitude = args[0]
            self.phase = math.fmod(math.radians(args[1]), math.pi)
            self.real = self.magnitude * cmath.cos(self.phase)
            self.imag = self.magnitude * cmath.sin(self.phase)
        else:
            self.magnitude = 0
            self.phase = 0
            self.real = 0
            self.imag = 0

    def __str__(self):
        return f"{self.magnitude:.3f}\u2220 {self.phase:.3f}\u00b0"
    
    def __add__(self, other: 'phasor') -> 'phasor':
        real = self.real + other.real
        imag = self.imag + other.imag
        return phasor(complex(real,imag))

    def __sub__(self, other: 'phasor') -> 'phasor':
        real = self.real - other.real
        imag = self.imag - other.imag
        return phasor(complex(real,imag))

    # TODO: Handle negative magnitudes
    def __mul__(self, other: 'phasor') -> 'phasor':
        magnitude = self.magnitude * other.magnitude
        phase = (self.phase + other.phase) % 360
        return phasor(magnitude, phase)

    def __truediv__(self, other: 'phasor') -> 'phasor':
        magnitude = self.magnitude / other.magnitude
        phase = (self.phase - other.phase) % 360
        return phasor(magnitude, phase)

    def __eq__(self, other: 'phasor') -> bool:
        if(cmath.isclose(self.magnitude, other.magnitude) and
                cmath.isclose(self.phase, other.phase)):
            return True
        else:
            return False

    def clear(self):
        self.magnitude = 0
        self.phase = 0
        self.real = 0
        self.imag = 0

    def set_magnitude(self, new_magnitude: float) -> None:
        self.magnitude = new_magnitude
        self.real = new_magnitude * math.cos(self.phase)
        self.imag = new_magnitude * math.sin(self.phase)

    def set_phase(self, new_phase: float) -> None:
        self.phase = new_phase
        self.real = self.magnitude * math.cos(new_phase)
        self.imag = self.magnitude * math.sin(new_phase)

    def get_magnitude(self) -> float:
        return self.magnitude

    def get_phase(self) -> float:
        return math.degrees(self.phase)

    def get_real(self) -> float:
        return self.real

    def get_imag(self) -> float:
        return self.imag

    def get_complex(self) -> complex:
        return (self.real+self.imag*1j)

def calc_mag(complex_num: complex) -> float:
    return abs(complex_num)

def calc_phase(complex_num: complex) -> float:
    return math.degrees(cmath.phase(complex_num))

def calc_real(num: phasor) -> complex:
    mag = num.get_magnitude()
    phase = num.get_phase()
    return mag*cmath.cos(phase)

def calc_imag(num: phasor) -> complex:
    mag = num.get_magnitude()
    phase = num.get_phase()
    return mag*cmath.sin(phase)
